deep-learning/src/main/scala/com/microsoft/azure/synapse/ml/onnx/ONNXUtils.scala

```diff
@@ -0,0 +1,275 @@
+import scala.reflect.ClassTag
+
+object ONNXUtils {
+  private[onnx] def mapOnnxJavaTypeToDataType(javaType: OnnxJavaType): DataType = {
+    javaType match {
+      case OnnxJavaType.INT8 => ByteType
+      case OnnxJavaType.INT16 => ShortType
+      case OnnxJavaType.INT32 => IntegerType
+      case OnnxJavaType.INT64 => LongType
+      case OnnxJavaType.FLOAT => FloatType
+      case OnnxJavaType.DOUBLE => DoubleType
+      case OnnxJavaType.BOOL => BooleanType
+      case OnnxJavaType.STRING => StringType
+      case OnnxJavaType.UNKNOWN => BinaryType
+    }
+  }
+
+  private[onnx] def mapTensorInfoToDataType(tensorInfo: TensorInfo): DataType = {
+    val dataType = mapOnnxJavaTypeToDataType(tensorInfo.`type`)
+
+    def nestedArrayType(depth: Int, dataType: DataType): ArrayType = {
+      if (depth == 1)
+        ArrayType(dataType)
+      else
+        ArrayType(nestedArrayType(depth - 1, dataType))
+    }
+
+    if (tensorInfo.isScalar) {
+      dataType
+    } else if (tensorInfo.getShape.length == 1) {
+      // first dimension is assumed to be batch size.
+      dataType
+    } else {
+      nestedArrayType(tensorInfo.getShape.length - 1, dataType)
+    }
+  }
+
+  @tailrec
+  private[onnx] def mapValueInfoToDataType(valueInfo: ValueInfo): DataType = {
+    valueInfo match {
+      case mapInfo: MapInfo =>
+        val keyDataType = mapOnnxJavaTypeToDataType(mapInfo.keyType)
+        val valueDataType = mapOnnxJavaTypeToDataType(mapInfo.valueType)
+        MapType(keyDataType, valueDataType)
+      case seqInfo: SequenceInfo =>
+        if (seqInfo.sequenceOfMaps) {
+          mapValueInfoToDataType(seqInfo.mapInfo)
+        } else {
+          mapOnnxJavaTypeToDataType(seqInfo.sequenceType)
+        }
+      case tensorInfo: TensorInfo =>
+        mapTensorInfoToDataType(tensorInfo)
+    }
+  }
+
+  private[onnx] def mapOnnxValueToArray(value: OnnxValue): Seq[Any] = {
+    value.getInfo match {
+      case tensorInfo: TensorInfo =>
+        if (tensorInfo.isScalar)
+          Seq(value.getValue)
+        else {
+          value.getValue.asInstanceOf[Array[_]].toSeq
+        }
+      case sequenceInfo: SequenceInfo =>
+        if (sequenceInfo.sequenceOfMaps) {
+          value.getValue.asInstanceOf[java.util.List[java.util.Map[_, _]]]
+            .asScala.toArray.map(_.asScala.toMap)
+        } else {
+          value.getValue.asInstanceOf[java.util.List[_]].asScala
+        }
+      case _: MapInfo =>
+        Array(value.getValue.asInstanceOf[java.util.Map[_, _]].asScala.toMap)
+    }
+  }
+
+  private[onnx] def createTensor(env: OrtEnvironment, tensorInfo: TensorInfo, batchedValues: Seq[_]): OnnxTensor = {
+    val shape: Array[Long] = tensorInfo.getShape
+    // the first dimension of the shape can be -1 when multiple inputs are allowed. Setting it to the real
+    // input size. Otherwise we cannot create the tensor from the 1D array buffer.
+    shape(0) = batchedValues.length
```

Comment:
Please add a check after this line:

```scala
    if (shape.tail.contains(-1))
      throw new Exception(s"The input tensor has shape [${shape.mkString}], " +
        s"but -1 is only allowed at the first dimension (batch size).")
```

This is for the case where the ONNX model expects a variable-length string array as input, where the input tensor shape would be [-1, -1] (the first -1 indicates unlimited batch size, and the second -1 indicates variable-length). Given the current setup, we cannot support this easily.